### ISSUE TITLE
[Question]: Test for custom dom events being emitted

### ISSUE THREAD

**Issue by adisekar** [microsoft/playwright#19546]
Hi, 
I am dispatching custom dom events, and is there a way to check if the event was fired?
Eg: Say I have a custom event for 'messageSent', I need a way to add a handler to make sure it is firing. 
Manually we can add a document.addEventListener() to check if callback was fired, but how can I test this? Is there any helpers or anyway I can add this listener to the page context to test these?
Thanks!

**Comment by adisekar** [comment]
Thanks @pavelfeldman
I was trying something similar to approach 1. But if the event does not occur, the promise times out and errors the test, is there a clean way to reject() the promise, if there is no event triggered?

**Comment by adisekar** [comment]
I am able to get approach 1 working for click() event, but it does not work for .focus() event. It seems the focus() event does not go through the event listener. Any reason why the focus() will behave differently than the click()?
Thank you!

**Comment by pavelfeldman** [comment]
No difference as far as Playwright is concerned, but it could be that your web app does different things for click and for focus, so it won't dispatch custom events on focus.

Instead of waiting, you can `addInitScript` that stores all the events in some array (window.myEvents) and then `page.evaluate(() => window.myEvents)` when you want to check them.

```js
await page.addInitScript(() => {
  window.fooEvents = [];
  document.addEventListener('foo', event => window.fooEvents.push(event.data);
});
```

```js
function takeFooEvents(page) {
  return page.evaluate(() => {
    const results = window.fooEvents;
    window.fooEvents = [];
    return results;
  });
}
```

Now all the events are always stored in array and you can retrieve / clear them when you need it in your test via `await takeFooEvents(page)`.

**Comment by adisekar** [comment]
Thanks @pavelfeldman 
However, I suspect that the issue is with .focus(), that it does not wait for event handlers to be attached, as I found in #[2902](https://github.com/microsoft/playwright/issues/2902) and [#14422](https://github.com/microsoft/playwright/issues/14422)

The click action waits for listeners to be attached but focus does not.  Adding this feature would be very helpful. 
They mentioned using waitForFunction in those issues, I am going to try if that can be a workaround

**Comment by pavelfeldman** [comment]
> The click action waits for listeners to be attached but focus does not.

That does not happen for click either. I'm not familiar with the frameworks that would be able to detect the "late hydration bug" where listeners are attached some time after the element becomes actionable and is interacted with.

**Comment by pavelfeldman** [comment]
Closing as per above, please feel free to open a new issue if this does not cover your use case.

**Comment by adisekar** [comment]
@pavelfeldman Thank you for your answers, it was very helpful. 
I was able to get the my issue resolved. Turns out the .focus() does not [autowait](https://playwright.dev/docs/actionability) for custom events to be bound, like the .click() does. So I had to add a waitForSelector(), and in my case to wait for hydration to happen. In our case we are using Stencil web components, to anyone else, who might find this useful.
